core/domain/auth_jobs_one_off.py

```diff
@@ -39,6 +39,258 @@
+        """Yields debug information for associations mapped to the same auth_id.
+
+        NOTE: Since assoc_by_auth_id and Firebase accounts are keyed by auth_id,
+        and are therefore guaranteed to have no collisions, the associations
+        reported will always come from UserAuthDetailsModel.
+
+        Args:
+            auth_id: str. The auth_id to check.
+            assoc_info_pairs: list(tuple(str, (str, bool))). The list of
+                associations that do not correspond to any auth_id.
+
+        Yields:
+            str. A debug string for the associations with collisions.
+        """
+        user_id_collisions = sorted(
+            '"%s"' % user_id for assoc_key, (user_id, _) in assoc_info_pairs
+            if assoc_key == cls.ASSOC_BY_USER_ID_KEY)
+
+        if len(user_id_collisions) > 1:
+            yield '%d UserAuthDetailsModels have auth_id="%s": %s' % (
+                len(user_id_collisions), auth_id, ', '.join(user_id_collisions))
+
+    @classmethod
+    def report_assoc_inconsistencies(
+            cls, auth_id, assoc_by_auth_id=None, assoc_by_user_id=None,
+            firebase_account_disabled=None):
+        """Reports inconsistencies between the given values.
+
+        IMPORTANT: The names of the keyword arguments MUST match the values of
+        their corresponding class keys: ASSOC_BY_AUTH_ID_KEY,
+        ASSOC_BY_USER_ID_KEY, FIREBASE_ACCOUNT_DISABLED_KEY.
+
+        Args:
+            auth_id: str. The auth_id of the user.
+            assoc_by_auth_id: tuple|None. The (user_id, deleted) properties from
+                the UserIdByFirebaseAuthIdModel corresponding to auth_id. NOTE:
+                UserIdByFirebaseAuthIdModel is keyed by auth_id,
+                so this argument will never use the default value of None.
+            assoc_by_user_id: tuple|None. The (user_id, deleted) properties from
+                the UserAuthDetailsModel corresponding to auth_id, or None if
+                one does not exist.
+            firebase_account_disabled: bool|None. Whether the Firebase account
+                corresponding to auth_id is disabled, or None if an account does
+                not exist.
+
+        Yields:
+            str. Debug information about a discovered inconsistency.
+        """
+        if assoc_by_auth_id is None and assoc_by_user_id is None:
+            # User is deleted and will be managed by resolve_user_conflicts().
+            return
+
+        distinct_user_ids = {assoc_by_auth_id[0]}
+        distinct_deleted_bools = {assoc_by_auth_id[1]}
+
+        if assoc_by_user_id is None:
+            yield (
+                'UserIdByFirebaseAuthIdModel(id="%s") does not correspond to a '
+                'unique UserAuthDetailsModel' % auth_id)
+        else:
+            distinct_user_ids.add(assoc_by_user_id[0])
+            distinct_deleted_bools.add(assoc_by_user_id[1])
+
+        if len(distinct_user_ids) == 2:
```

Comment:
Rather than make sets, I think it would be clearer to just have "if X != Y"

Reply:
I considered that, but the names of X and Y end up being a bit too complicated/unclear, so I went with this at the end. How strongly do you feel about it?

Reply:
Moderately strongly. It took me a while to puzzle out what was going on here.

Maybe something like auth_details_model_user_id, auth_details_model_deleted_bool, firebase_auth_model_user_id, firebase_auth_model_deleted_bool?

Reply:
Done